### PR TITLE
Implemented wildcards in unifications and disequality constraints

### PR DESCRIPTION
**Pull request by Damtev**
`TODO`

### CODE REVIEW

File: klogic-core/src/main/kotlin/org/klogic/core/Term.kt [hunk 1]
```diff
@@ -120,11 +120,17 @@ sealed interface Term<T : Term<T>> {
     }
 }
 
+/**
+ * Represents a logic term of the specific type
+ * that does not have a particular value before unifications - either [Var] or [Wildcard].
+ */
+sealed interface UnboundedValue<T : Term<T>> : Term<T>
```
Comment (Kakadu) [inline klogic-core/src/main/kotlin/org/klogic/core/Term.kt:127]: See my remark about single wildcard

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            val run = run(10, goal)
+            val answer = run.single()
+
+            assertEquals(one, answer.term.asReified().first)
+
+            val wildcardInPair = answer.term.asReified().second as Wildcard<Symbol>
+            val answerConstraint = answer.constraints.single() as InequalityConstraint
+
+            val expectedConstraint = SingleInequalityConstraint(wildcardInPair, one)
+            assertTrue(answerConstraint.simplifiedConstraints.any { it == expectedConstraint })
```
Comment (Kakadu) [inline klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:74]: Я думаю тут не должно быть ответов

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            val a = (-1).createTypedVar<Symbol>()
+            val b = (-2).createTypedVar<Symbol>()
+            val goal = { x: Term<Symbol>, y: Term<Symbol> -> (x `===` freshTypedWildcard()) and (y `===` freshTypedWildcard()) }
+
+            val run = unreifiedRun(10, goal(a, b))
+            val answers = listOf(a, b).map { run.reify(it) }
+
+            assertEquals(a, answers[0].single().term)
+            assertEquals(b, answers[1].single().term)
+        }
```
Comment (Kakadu) [inline klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:53]: В OCanren  не задумывалась унификаций вайлдкардов с переменными. Какая должна быть семантика -- я хз, скорее всего все вайлдкарды должны считаться свежими переменными.

File: klogic-core/src/main/kotlin/org/klogic/core/Context.kt [hunk 4]
```diff
@@ -21,10 +21,15 @@ open class RelationalContext : AutoCloseable {
...
     /**
-     * The index of the last variable created in this context.
+     * The index of the last [Var] created in this context.
      */
     private var lastCreatedVariableIndex: Int = 0
 
+    /**
+     * The index of the last [Wildcard] created in this context.
+     */
+    private var lastCreatedWildcardIndex: Int = 0
```
[the remark this answers is not in the document]
Comment (Damtev) [reply klogic-core/src/main/kotlin/org/klogic/core/Context.kt:31]: In such a case, what is this variable with the index 11 [here](https://github.com/PLTools/OCanren/compare/master...Kakadu:OCanren:wildcards+var_ordering#diff-bffd450de54f46a22f1a0ad4f65477865577d9156196aae6caf4ad5a4eff9105R13-R15)?

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt (hunk 2 above)
Comment (Damtev) [reply klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:74]: Once again, `OCanren` [thinks they should be](https://github.com/PLTools/OCanren/compare/master...Kakadu:OCanren:wildcards+var_ordering#diff-bffd450de54f46a22f1a0ad4f65477865577d9156196aae6caf4ad5a4eff9105R13-R15)

File: klogic-core/src/main/kotlin/org/klogic/core/Context.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kakadu) [reply klogic-core/src/main/kotlin/org/klogic/core/Context.kt:31]: link is broken

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt (hunk 2 above)
Comment (Kakadu) [reply klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:74]: link is broken

File: klogic-core/src/main/kotlin/org/klogic/core/Context.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Damtev) [reply klogic-core/src/main/kotlin/org/klogic/core/Context.kt:31]: Fixed

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt (hunk 2 above)
Comment (Damtev) [reply klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:74]: Fixed

File: klogic-core/src/main/kotlin/org/klogic/core/Context.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kakadu) [reply klogic-core/src/main/kotlin/org/klogic/core/Context.kt:31]: > In such a case, what is this variable with the index 11 [here](https://github.com/Kakadu/OCanren/blob/4bcc41720496999166a0d823dcec7e9540dc0c8f/test_wc/main.t#L13-L15)?

A fresh variable with index 11

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt (hunk 2 above)
Comment (Kakadu) [reply klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:74]: In the answer you have explicit wildcards. OCanren test doesn't have wildcards in the answer

**Comment by Kakadu** [comment]
It would be great to add canonical examples, for example, "a peano number is less than 4"

**Review by Kakadu — APPROVED** [review]

File: klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            val value = "42".toSymbol()
+
+            fun allLists(x: Term<LogicList<Symbol>>): Goal = conde(
+                x `===` emptyLogicList(),
+                freshTypedVars<LogicList<Symbol>> { prev ->
+                    (x `===` Cons(value, prev)) and delay { allLists(prev) }
+                }
+            )
+
+            fun isNotCons(x: Term<LogicList<Symbol>>) = x `!==` Cons(freshTypedWildcard(), freshTypedWildcard())
```
Comment (Kakadu) [inline klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt:160]: It is rather complicated name, maybe just `wildcard()`?

Comment (Damtev) [reply klogic-core/src/test/kotlin/org/klogic/core/wildcards/WildcardsTest.kt]: Done

**Pull request merged**